examples/cis2-wccd-upgrade/src/lib.rs

```diff
@@ -0,0 +1,2030 @@
+//! upgradable wCCD token
+
+#![cfg_attr(not(feature = "std"), no_std)]
+use crate::CustomContractError::{
+    AlreadyInitialized, ContractPaused, OnlyImplementation, OnlyState, OutOfBound, StateInvokeError,
+};
+use concordium_cis2::*;
+use concordium_std::*;
+use core::fmt::Debug;
+
+/// The id of the wCCD token in this contract.
+const TOKEN_ID_WCCD: ContractTokenId = TokenIdUnit();
+
+/// The metadata url for the wCCD token.
+const TOKEN_METADATA_URL: &str = "https://some.example/token/wccd";
+
+// Types
+
+/// Contract token ID type.
+/// Since this contract will only ever contain this one token type, we use the
+/// smallest possible token ID.
+type ContractTokenId = TokenIdUnit;
+
+/// Contract token amount type.
+/// Since this contract is wrapping the CCD and the CCD can be represented as a
+/// u64, we can specialize the token amount to u64 reducing module size and cost
+/// of arithmetics.
+type ContractTokenAmount = TokenAmountU64;
+
+/// The state tracked for each address.
+#[derive(Serial, DeserialWithState, Deletable)]
+#[concordium(state_parameter = "S")]
+struct AddressState<S> {
+    /// The number of tokens owned by this address.
+    balance:   ContractTokenAmount,
+    /// The address which are currently enabled as operators for this token and
+    /// this address.
+    operators: StateSet<Address, S>,
+}
+
+/// The contract state.
+#[derive(Serial, DeserialWithState)]
+#[concordium(state_parameter = "S")]
+struct State<S> {
+    /// Address of the w_ccd proxy contract.
+    proxy_address:          Option<ContractAddress>,
+    /// Address of the w_ccd implementation contract.
+    implementation_address: Option<ContractAddress>,
+    /// FALSE if proxy and implementation addresses have not been set yet.
+    /// TRUE if proxy and implementation addresses have been set.
+    initialized:            bool,
+    /// The state of the one token.
+    token:                  StateMap<Address, AddressState<S>, S>,
+    /// Timestamp when contract is unpaused.
+    unpause_time:           Timestamp,
+}
```

Comment:
I think something like this would make more sense 😊 
```
enum State<S> {
    Uninitialized,
    Initialized {
      proxy_address: ContractAddress,
      implementation_address: ContractAddress,
      token: StateMap<Address, AddressState<S>, S>,
      unpause_time: Timestamp,
     }
}

Reply:
The same idea can be applied to `StateImplementation` the proxy state.